File: businessdb.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "hashtable.h"

int main() {
	HashTable table;
	init_table(&table);
	load_database(&table, "customers.tsv");

	char command[256];
	while (1) {
    	printf("Enter command: ");
    	scanf("%s", command);

    	if (strcmp(command, "add") == 0) {
        	char email[256], name[256], food[256], shoe_size_str[256];
        	int shoe_size;
        	char *endptr;
        	printf("Enter email: ");
        	scanf("%s", email);
        	printf("Enter name: ");
        	scanf("%s", name);
    		while (1) {
       		 printf("Enter shoe size: ");
       		 scanf("%s", shoe_size_str);
   	 	 shoe_size = strtol(shoe_size_str, &endptr, 10);
   	 	 if(*endptr == '\0' && endptr != shoe_size_str){
   		  break;
   		 }
   	 	 else{
   		  printf("Invalid shoe size\n");
   	 	 }
       	       }
               printf("Enter favorite food: ");
               scanf("%s", food);
               add_customer(&table, email, name, shoe_size, food);
    	} 
	else if (strcmp(command, "lookup") == 0) {
        	char email[256];
        	printf("Enter email: ");
        	scanf("%s", email);
        	Customer *customer = lookup_customer(&table, email);
        	if (customer) {
            	printf("Email: %s, Name: %s, Shoe Size: %d, Favorite Food: %s\n",
                   	customer->email, customer->name, customer->shoe_size,
                   	customer->favorite_food);
        	} 
		else {
            		printf("Customer not found!\n");
        	}
    	} 
	else if (strcmp(command, "delete") == 0) {
        	char email[256];
        	printf("Enter email: ");
        	scanf("%s", email);
        	delete_customer(&table, email);
    	} 
	else if (strcmp(command, "list") == 0) {
        	list_customers(&table);
    	} 
	else if (strcmp(command, "save") == 0) {
        	save_database(&table, "customers.tsv");
    	} 
	else if (strcmp(command, "quit") == 0) {
        	break;
    	} 
	else {
        	printf("Unknown command!\n");
    	}
	}

	free_table(&table);
	return 0;
}

File: businessdb.py
#!/usr/bin/env python3

class Customer:
    def __init__(self, email, name, shoesize, food):
        self.email = email
        self.name = name
        self.shoesize = shoesize
        self.food = food
    def __str__(self):
        return ("email: {}\nname: {}\nshoesize: {}\nfood: {}\n".format(
            self.email, self.name, self.shoesize, self.food))

def main():
    customers = {}

    with open("customers.tsv") as infile:
        # how to do this in C? fgets
        for line in infile:
            # remove the newline in C?
            line = line.strip()
            # how to find tab characters in C?
            # maybe strchr?
            fields = line.split('\t')
            email = fields[0]
            name = fields[1]
            shoesize = int(fields[2])
            food = fields[3]
            customer = Customer(email, name, shoesize, food)

            customers[email] = customer

    while True:
        command = input("command: ")

        if command == "quit":
            break
        elif command == "save":
            # how to do this in C?
            # open it with fopen in C (for writing here)
            with open("customers.tsv", "w") as outfile:
                for email,customer in customers.items():
                    line = "{}\t{}\t{}\t{}".format(customer.email,
                                                   customer.name,
                                                   customer.shoesize,
                                                   customer.food)
                    print(line, file=outfile)
            # remember to fclose in C
        elif command == "add":
            # Note that in your program, you're going to have to consider the
            # possibility that the user might type a tab character here; you
            # should complain if they do.
            email = input("email address? ")
            name = input("name? ")
            shoesize = int(input("shoe size? "))
            food = input("favorite food? ")
            customer = Customer(email, name, shoesize, food)
            customers[email] = customer
        elif command == "lookup":
            email = input("email address? ")
            if email in customers:
                print(customers[email])
            else:
                print("user not found!")
        elif command == "delete":
            email = input("email address? ")
            if email in customers:
                del customers[email]
            else:
                print("user not found!")
        elif command == "list":
            for email,customer in customers.items():
                print(customer)
        else:
            print("unknown command")

if __name__ == "__main__": main()
